BasicSampleKotlin/app/src/main/java/com/example/android/persistence/db/AppDatabase.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2017, The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.persistence.db
+
+import android.arch.lifecycle.LiveData
+import android.arch.lifecycle.MutableLiveData
+import android.arch.persistence.db.SupportSQLiteDatabase
+import android.arch.persistence.room.Database
+import android.arch.persistence.room.Room
+import android.arch.persistence.room.RoomDatabase
+import android.arch.persistence.room.TypeConverters
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+
+import com.example.android.persistence.AppExecutors
+import com.example.android.persistence.db.converter.DateConverter
+import com.example.android.persistence.db.dao.CommentDao
+import com.example.android.persistence.db.dao.ProductDao
+import com.example.android.persistence.db.entity.CommentEntity
+import com.example.android.persistence.db.entity.ProductEntity
+
+@Database(entities = arrayOf(ProductEntity::class, CommentEntity::class), version = 1)
+@TypeConverters(DateConverter::class)
+abstract class AppDatabase : RoomDatabase() {
+
+    private val isDatabaseCreated = MutableLiveData<Boolean>()
+
+    val databaseCreated: LiveData<Boolean>
+        get() = isDatabaseCreated
+
+    abstract fun productDao(): ProductDao
+
+    abstract fun commentDao(): CommentDao
+
+    /**
+     * Check whether the database already exists and expose it via [.getDatabaseCreated]
+     */
+    private fun updateDatabaseCreated(context: Context) {
+        if (context.getDatabasePath(DATABASE_NAME).exists()) {
+            setDatabaseCreated()
+        }
+    }
+
+    private fun setDatabaseCreated() {
+        isDatabaseCreated.postValue(true)
+    }
+
+    companion object {
+        private var instance: AppDatabase? = null
+
+        @VisibleForTesting
+        const val DATABASE_NAME = "basic-sample-db"
+
+        @JvmStatic
+        fun getInstance(context: Context, executors: AppExecutors): AppDatabase {
+            if (instance == null) {
+                synchronized(AppDatabase::class.java) {
+                    if (instance == null) {
+                        instance = buildDatabase(context.applicationContext, executors).also { db ->
+                            db.updateDatabaseCreated(context.applicationContext)
```

Comment:
It seems like we set database created twice (once in the callback on the builder and then again when we get the database back) - do you know why we need both?

Reply:
It seems to me that one is just a sanity check if the DB already existed on start-up, while the other one is actually tracking that the DB was just created.

I assume this is all tied into a big LiveData for some reason to hide the initial "empty list" calls.

Not sure if all this dance is necessary though, I mean the LiveData would be triggered and would update the view with the new and inserted initial data anyway.